java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerSessionMiddleware.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Middleware for handling Flight SQL Sessions including session cookie handling.
+ *
+ * Currently experimental.
+ */
+public class ServerSessionMiddleware implements FlightServerMiddleware {
+  Factory factory;
+  boolean existingSession;
+  private Session session;
+  private String closedSessionId = null;
+
+  public static final String sessionCookieName = "arrow_flight_session_id";
+
+  /**
+   * Factory for managing and accessing ServerSessionMiddleware.
+   */
+  public static class Factory implements FlightServerMiddleware.Factory<ServerSessionMiddleware> {
+    private final Map<String, Session> sessionStore =
+        new ConcurrentHashMap<>();
+    private final Callable<String> idGenerator;
+
+    /**
+     * Construct a factory for ServerSessionMiddleware.
+     *
+     * Factory manages and accesses persistent sessions based on HTTP cookies.
+     *
+     * @param idGenerator A Callable returning unique session id Strings.
+     */
+    public Factory(Callable<String> idGenerator) {
+      this.idGenerator = idGenerator;
+    }
+
+    private synchronized Session createNewSession() {
+      String id;
+      try {
+        id = idGenerator.call();
+      } catch (Exception ignored) {
+        // Most impls aren't going to throw so don't make caller handle a nonexistent checked exception
+        return null;
+      }
+      if (sessionStore.containsKey(id)) {
+        // Collision, should also never happen
+        return null;
```

Comment:
Throw an AssertionError or some other thing to indicate a violated invariant

Reply:
Making both of these throw CallStatus.INTERNAL